src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1989,6 +2001,51 @@ def make_api_call_with_retries(uri, data, tls_verify, fault_type, summary, cli_e
                 close_subprocess_and_raise_cli_error(clientproxy_process, cli_error + str(e))
 
 
+def fetch_pop_publickey_kid(api_server_port, clientproxy_process):
+    requestbody = {}
+    poppublickey_uri = f'https://localhost:{api_server_port}/identity/poppublickey'
+    # Needed to prevent skip tls warning from printing to the console
+    original_stderr = sys.stderr
+    f = open(os.devnull, 'w')
+    sys.stderr = f
+
+    get_publickey_response = make_api_call_with_retries(poppublickey_uri, requestbody, "get", False, consts.Get_PublicKey_Info_Fault_Type,
+                                                        'Unable to fetch public key info from clientproxy',
+                                                        "Failed to fetch public key info from client proxy", clientproxy_process)
+
+    sys.stderr = original_stderr
+    publickey_info = json.loads(get_publickey_response.text)
+    kid = publickey_info['publicKey']['kid']
+
+    return kid
+
+
+def fetch_and_post_at_to_csp(cmd, api_server_port, tenantId, kid, clientproxy_process):
+    req_cnfJSON = {"kid": kid, "xms_ksl": "sw"}
+    req_cnf = base64.urlsafe_b64encode(json.dumps(req_cnfJSON).encode('utf-8')).decode('utf-8')
+
+    # remove padding '=' character
+    if req_cnf[len(req_cnf) - 1] == '=':
+        req_cnf = req_cnf[:-1]
+
+    token_data = {"token_type": "pop", "key_id": kid, "req_cnf": req_cnf}
+    profile = Profile(cli_ctx=cmd.cli_ctx)
+    credential, _, _ = profile.get_login_credentials(subscription_id=profile.get_subscription()["id"], resource=consts.KAP_1P_Server_App_Scope)
```

Comment:
what are other two return objects? Can this method throw exception?

Reply:
subID and tenantID are the two other objects. get_login_creds won't throw an exception but it's a good idea to capsule the AT fetch part in try catch . Thanks for pointing out. Updated it to handle any exceptions